src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.cs

```diff
@@ -110,121 +135,156 @@ public static SyntaxNode LocateOwner(this SyntaxNode node, SourceChange change)
         {
-            var child = children[i];
-            owner = LocateOwner(child, change);
-            if (owner != null)
+            foreach (var child in list)
             {
-                break;
+                if (child.LocateOwner(change) is { } owner)
+                {
+                    return owner;
+                }
             }
-        }
 
-        return owner;
+            return null;
+        }
     }
 
     public static bool IsTransitionSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return TransitionSpanKinds.Contains(node.Kind);
+        return s_transitionSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsMetaCodeSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return MetaCodeSpanKinds.Contains(node.Kind);
+        return s_metaCodeSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsCommentSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return CommentSpanKinds.Contains(node.Kind);
+        return s_commentSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsCodeSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return CodeSpanKinds.Contains(node.Kind);
+        return s_codeSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsMarkupSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return MarkupSpanKinds.Contains(node.Kind);
+        return s_markupSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsNoneSpanKind(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
-        return NoneSpanKinds.Contains(node.Kind);
+        return s_noneSpanKinds.Contains(node.Kind);
     }
 
     public static bool IsSpanKind(this SyntaxNode node)
+        => s_allSpanKinds.Contains(node.Kind);
+
+    private static IEnumerable<SyntaxNode> FlattenSpansInReverse(this SyntaxNode node)
     {
-        return IsTransitionSpanKind(node) ||
-            IsMetaCodeSpanKind(node) ||
-            IsCommentSpanKind(node) ||
-            IsCodeSpanKind(node) ||
-            IsMarkupSpanKind(node) ||
-            IsNoneSpanKind(node);
+        using var stack = new NodeStack(node.DescendantNodes());
```

Comment:
Using a thread-affinitized collection in an iterator strikes me as a latent bug waiting to strike. I'd prefer to just use `ArrayBuilder` from the pooled objects types here.

Reply:
I'll await your PR to bring in immutable collections so I can use a pool of `ImmutableArray<T>.Builder`.

Reply:
Actually, I switched over to an `ObjectPool<SyntaxNode[]>`.